src/Orleans/Hosting/ServiceProviderBuilder.cs

```diff
@@ -0,0 +1,81 @@
+using System;
+using System.Collections.Generic;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Orleans.Hosting
+{
+    /// <summary>
+    /// Functionality for building <see cref="IServiceProvider"/> instances.
+    /// </summary>
+    internal class ServiceProviderBuilder
+    {
+        private readonly List<Action<IServiceCollection>> configureServicesDelegates = new List<Action<IServiceCollection>>();
+        private readonly List<object> configureContainerDelegates = new List<object>();
+        private IServiceProviderFactoryAdapter serviceProviderFactory;
+
+        /// <summary>
+        /// Builds the service provider.
+        /// </summary>
+        /// <returns>The service provider.</returns>
+        public IServiceProvider BuildServiceProvider()
+        {
+            // Configure the container.
+            var services = new ServiceCollection();
+            foreach (var configureServices in this.configureServicesDelegates)
+            {
+                configureServices(services);
+            }
+
+            // If no service provider factory has been specified, set a default.
+            if (this.serviceProviderFactory == null)
+            {
+                var factory = new DelegateServiceProviderFactory(svc => svc.BuildServiceProvider());
+                this.UseServiceProviderFactory(factory);
```

Comment:
This default assignment should be done in a separate method that doesn't get inlined, so that it doesn't even attempt to load that type. Probably not a big deal, since the assembly will already be loaded, as we use some extension methods from the non-abstractions library elsewhere

Reply:
I moved it into a method marked as `NoInlining`